### PR TITLE
[WIP] template class for isl_*_list + iterators

### PR DESCRIPTION
**Pull request by ftynse**
This is how a template class for the isl lists could look like.

The problem with iterators is that we can hardly provide standard-compliant iterators as ForwardIterator concept requires to return references to elements (http://en.cppreference.com/w/cpp/concept/ForwardIterator), which would contradict the design of "by-value-only".  Anyway, I tried to make it as close as possible.  Range-based for will work, but some standard algorithms assume certain iterator categories.

@tobig @chandangreddy

### CODE REVIEW

**Comment by ftynse** [comment]
I checked, and most standard algorithms require the InputIterator concept, which we can support.  We cannot provide a standard-compliant RandomAccessIterator (because of that reference problem, which requires the container class to actually have access to the data store), but nobody prevents us from making our own.

Also, the requirement for reference typedef to be an actual reference type rather than some writable-to proxy seems rather constraining.  I would check in the recent standard.

We also can provide a back_inserter style OutputIterator, but not a regular vector-style OutputIterator through which one could write as it will modify the underlying object rather than return a new one.

(On a side note, don't forget to mention somewhere that isl++ containers are not compatible with C++17 parallel algorithms, which incidentally also require ForwardIterator).

**Comment by tobig** [comment]
Very interesting.

File: interface/test/test_islpp.cc [hunk 1]
```diff
@@ -383,6 +384,33 @@ TEST(ISLPP, IdLifetime) {
...
+  isl::ctx ctx(isl_ctx_alloc());
+  ScopeGuard g([&]() { isl_ctx_free(ctx.release()); });
+
+  isl_id_list *c_list = isl_id_list_alloc(ctx.get(), 2);
+  isl::id id1(ctx, std::string("id1")), id2(ctx, std::string("id2"));
+  isl::id id1c(id1), id2c(id2);
+  isl_id *id3 = isl_id_alloc(ctx.get(), "foo", NULL);
+  isl::id id3c(isl_id_copy(id3));
+  c_list = isl_id_list_add(c_list, id1.release());
+  c_list = isl_id_list_add(c_list, id2.release());
```
Comment (tobig) [inline interface/test/test_islpp.cc]: Why this explicit release. You don't really optimize for performance here, no?

File: interface/test/test_islpp.cc [hunk 2]
```diff
@@ -383,6 +384,33 @@ TEST(ISLPP, IdLifetime) {
...
 
+TEST(ISLPP, IdList) {
+  isl::ctx ctx(isl_ctx_alloc());
+  ScopeGuard g([&]() { isl_ctx_free(ctx.release()); });
+
+  isl_id_list *c_list = isl_id_list_alloc(ctx.get(), 2);
+  isl::id id1(ctx, std::string("id1")), id2(ctx, std::string("id2"));
+  isl::id id1c(id1), id2c(id2);
+  isl_id *id3 = isl_id_alloc(ctx.get(), "foo", NULL);
+  isl::id id3c(isl_id_copy(id3));
```
Comment (tobig) [inline interface/test/test_islpp.cc]: Why these explicit copies?

File: interface/test/test_islpp.cc [hunk 3]
```diff
@@ -383,6 +384,33 @@ TEST(ISLPP, IdLifetime) {
   EXPECT_EQ(0, *flag);
 }
 
+TEST(ISLPP, IdList) {
+  isl::ctx ctx(isl_ctx_alloc());
+  ScopeGuard g([&]() { isl_ctx_free(ctx.release()); });
+
+  isl_id_list *c_list = isl_id_list_alloc(ctx.get(), 2);
+  isl::id id1(ctx, std::string("id1")), id2(ctx, std::string("id2"));
```
Comment (tobig) [inline interface/test/test_islpp.cc]: Why do you use the C interface to setup the list? It would be a lot nicer to do this in the C++ interface.

In fact, if we don't have the relevant functions in the C++ interface, should we not yet expose them first?

File: interface/isl.h.top [hunk 4]
```diff
@@ -251,5 +253,319 @@ enum class dim {
...
+}
+
+template <typename T>
+typename isl_list_traits<T>::list_type manage(T t) {
+  return typename isl_list_traits<T>::list_type(t);
+}
+
 }
 } // namespace isl
+
```
Comment (tobig) [inline interface/isl.h.top]: Unrelated.

File: interface/test/test_islpp.cc [hunk 5]
```diff
@@ -383,6 +384,33 @@ TEST(ISLPP, IdLifetime) {
   EXPECT_EQ(0, *flag);
 }
 
+TEST(ISLPP, IdList) {
```
Comment (tobig) [inline interface/test/test_islpp.cc]: Using gtest here will make it harder to merge these changes. I see little benefit in using a mix of gtest and and the isl tests for the C++ bindings.

File: interface/isl.h.top [hunk 6]
```diff
@@ -251,5 +253,319 @@ enum class dim {
   all = isl_dim_all
 };
 
+template <typename T>
+struct isl_list;
+
+class id;
+class basic_set;
+
+id manage(__isl_take isl_id*);
```
Comment (tobig) [inline interface/isl.h.top]: isl::id is not yet exported in isl. isl::val is exported and would make a similar good test case.

File: interface/isl.h.top [hunk 7]
```diff
@@ -251,5 +253,319 @@ enum class dim {
...
+
+template <>
+struct isl_list_traits<isl_id_list *> {
+  typedef isl_list<isl::id> list_type;
+};
+
+template <>
+struct isl_list_traits<isl_basic_set_list *> {
+  typedef isl_list<isl::basic_set> list_type;
+};
```
Comment (tobig) [inline interface/isl.h.top]: Why a basic set list as well. It is not tested and likely follows the very same code path. Removing it would make the code shorter. Or do you want to show that it works with different types. In this case, we should likely test both.

**Comment by ftynse** [comment]
My current understanding is that isl lists, like all other isl types, are essentially immutable.  Modifying a list returns a new list.  This is not really compatible with standard containers.  If the user wants a vector-like behavior, they can just construct a vector from a pair of iterators and use it.

File: interface/test/test_islpp.cc (hunk 3 above)
Comment (ftynse) [inline interface/test/test_islpp.cc]: That's the trade-off.  If we want immutable lists, we cannot construct them otherwise than from C lists.  We can, however, have STL-style construction from a pair of iterators: the list can be constructed but not modified.

I don't feel really confident about exposing an `isl_list isl_list::add(elem) const` as it makes too easy to (a) make copies of lists that are expensive (cow won't help, we are writing) and (b) mislead the user into expecting a mutable list.

**Comment by aisoard** [comment]
> Would be interesting to get input form other C++ experts. Alex, what do you think @aisoard?

I agree, it would be interesting to get input from other C++ experts! I will try to fake it.

Oh god... Do we need to support `isl_*_list`? I think Alex is right, we are quite limited in the kind of operators we can provide, although, I am ok with having a reference that is not actually a reference (but maybe I don't foresee the associated problems) as long as it behaves as expected. That being said, I am surprised T itself is sufficient, I expected the need to keep around a reference to the list wrapper (so that if we write, we can update the pointer to the C list inside it).

Also, I am guessing that's irrelevant at this stage, but why `isl::isl_list<>` instead of simply `isl::list<>`?

**Comment by ftynse** [comment]
On 10/08/17 13:05, Alexandre Isoard wrote:
>
>     Would be interesting to get input form other C++ experts. Alex,
>     what do you think @aisoard <https://github.com/aisoard>?
>
> I agree, it would be interesting to get input from other C++ experts! 
> I will try to fake it.
>
> Oh god... Do we need to support |isl_*_list|?
>
We have a couple of use cases with AST node manipulation where a list is 
returned and the C API exposes a list.  Alternative solution may be to 
change that API instead.
Anyway, there are things that are implicitly iterable (e.g. union_maps 
and maps) and we may want iterators for them, too.  So having something 
more trivially iterable will probably help us.  My first though was to 
just be able to construct iterators from C isl_*_list (which is also 
feasible), but then we would occasionally have raw C pointers in C++ 
code (or have to change the API that returns a list so that it returns a 
pair of iterators without exposing the underlying data structure).

> I think Alex is right, we are quite limited in the kind of operators 
> we can provide, although, I am ok with having a reference that is not 
> actually a reference (but maybe I don't foresee the associated 
> problems) as long as it behaves as expected.
>
So it should only be used where an InputIterator is expected. I would 
also go with a solution that has less code for us to support, and a 
BidirectionalIterator requires quite a few member functions.

> That being said, I am surprised T itself is sufficient, I expected the 
> need to keep around a reference to the list wrapper (so that if we 
> write, we can update the pointer to the C list inside it).
>
We do not intend to update the list through iterators.  This would make 
the isl_list type mutable and contradict the general semantics of the 
interface.

> Also, I am guessing that's irrelevant at this stage, but why 
> |isl::isl_list<>| instead of simply |isl::list<>|?
>
No particular reason.  My first draft was outside the namespace, so it 
stayed that way.

**Comment by ftynse** [comment]
Okay, so here's a way to construct an immutable isl_list from a any standard container (or even array) through a pair of iterators.

Looking at the growing list of overloads in the `detail` namespace, I am starting to think that we could just specialize the `isl_list` template for the relevant types in the generator.
Pros:
- can prevent users from attempting `isl::isl_list<int>` that would fail with a long list of errors due to failed overload resolution;
- better compilation time because no overload resolution;

Cons:
- less trivial code should be generated, which may be hard to check;
- if we want to extend the list interface with extra methods, potential code duplication.

Iterator can still be written as a template in terms of functions of template argument.

**Comment by ftynse** [comment]
Superseded by #27, closing.
